### DISCUSSION TITLE
How to document a fully dynamic API ?

### DISCUSSION THREAD

**Discussion by moweerkat** [OAI/OpenAPI-Specification#2946] [Q&A]
Hello,

I'm writing a platform that allows user to build their API. Some endpoints are "statics", most of them are really dynamic and we can't know in advance what one endpoint looks like, how it behaves. Requests, validations, responses quite likely depend on what the users set up.

Any suggestions, documentation, maybe third parties that make it work this way ?

Thanks

**Comment by ptrthomas** [comment]
@glamothe9 one possible solution is to write a bunch of tests that fleshes out the API behavior. happy to hear ideas on how to improve this demo: https://github.com/ptrthomas/karate-oas-demo

**Comment by flaksp** [comment]
OpenAPI document is just huge JSON file. You may simply generate separate OpenAPI JSON files for each user. When they will create endpoint in their account, your application also will add entry to OpenAPI document associated with that user.

It's not possible to write only one OpenAPI document that will cover all users in your case.

**Comment by handrews** [comment]
@moweerkat if you're still reading after nearly two years, OpenAPI at some point made a conscious decision to be a design-time, rather than run-time description.  For runtime description, you'd be better off looking at hypermedia tools.  Several formats are supported by [ketting](https://github.com/badgateway/ketting/), which out of the few hypermedia tools I've seen seems to be fairly mature with support for multiple formats.

**Comment by handrews** [comment]
I think this has been pretty thoroughly answered, so I'm going to close it.  Feel free to open a new discussion if needed.
